src/app/ReadClient.cpp

```
@@ -0,0 +1,249 @@
+    return "N/A";
+}
+
+void ReadClient::MoveToState(const ClientState aTargetState)
+{
+    mState = aTargetState;
+    ChipLogDetail(DataManagement, "Client[%u] moving to [%5.5s]",
+                  InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this), GetStateStr());
+}
+
+void ReadClient::ClearState(void)
+{
+    MoveToState(ClientState::Uninitialized);
+}
+
+CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdminId)
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    System::PacketBufferHandle msgBuf     = nullptr;
+    EventPathParams * eventPathParamsList = nullptr;
+    size_t aEventPathParamsListSize       = 0;
+    chip::EventNumber eventNumber         = 0;
+    ChipLogDetail(DataManagement, "Client[%u] [%5.5s]", InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this),
+                  GetStateStr());
+    VerifyOrExit(ClientState::Initialized == mState, err = CHIP_ERROR_INCORRECT_STATE);
+
+    if (nullptr != mpDelegate)
+    {
+        mpDelegate->HandlePathPrepareNeededCallBack(eventPathParamsList, aEventPathParamsListSize, eventNumber);
+    }
+
+    {
+        chip::System::PacketBufferTLVWriter writer;
+        ReadRequest::Builder request;
+
+        msgBuf = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLength);
+        VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+        writer.Init(std::move(msgBuf));
+
+        err = request.Init(&writer);
+        SuccessOrExit(err);
+
+        if (aEventPathParamsListSize != 0 && eventPathParamsList != nullptr)
+        {
+            // fill to construct event paths
+        }
+        request.EndOfReadRequest();
+        SuccessOrExit(request.GetError());
+
+        err = writer.Finalize(&msgBuf);
+        SuccessOrExit(err);
+
+        VerifyOrExit(msgBuf->EnsureReservedSize(System::PacketBuffer::kDefaultHeaderReserve), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    }
+
+    ClearExistingExchangeContext();
```

This seems like an odd thing to do.  I'd think trying to send a read on a read client that has an existing exchange would fail, instead of silently dropping the existing exchange...

---

yep, agree, fixed, thanks